L1Trigger/TrackFindingTMTT/src/TP.cc

```diff
@@ -0,0 +1,189 @@
+        charge_(tpPtr->charge()),
+        mass_(tpPtr->mass()),
+        pt_(tpPtr->pt()),
+        eta_(tpPtr->eta()),
+        theta_(tpPtr->theta()),
+        tanLambda_(1. / tan(theta_)),
+        phi0_(tpPtr->phi()),
+        vx_(tpPtr->vertex().x()),
+        vy_(tpPtr->vertex().y()),
+        vz_(tpPtr->vertex().z()),
+        d0_(vx_ * sin(phi0_) - vy_ * cos(phi0_)),                      // Copied from CMSSW class TrackBase::d0().
+        z0_(vz_ - (vx_ * cos(phi0_) + vy_ * sin(phi0_)) * tanLambda_)  // Copied from CMSSW class TrackBase::dz().
+  {
+    const vector<SimTrack>& vst = tpPtr->g4Tracks();
+    EncodedEventId eid = vst.at(0).eventId();
+    inTimeBx_ = (eid.bunchCrossing() == 0);  // TP from in-time or out-of-time Bx.
+    physicsCollision_ = (eid.event() == 0);  // TP from physics collision or from pileup.
+
+    this->fillUse();        // Fill use_ flag, indicating if TP is worth keeping.
+    this->fillUseForEff();  // Fill useForEff_ flag, indicating if TP is good for tracking efficiency measurement.
+  }
+
+  //=== Fill truth info with association from tracking particle to stubs.
+
+  void TP::fillTruth(const vector<Stub>& vStubs) {
+    for (const Stub& s : vStubs) {
+      for (const TP* tp_i : s.assocTPs()) {
+        if (tp_i->index() == this->index())
+          assocStubs_.push_back(&s);
+      }
+    }
+
+    this->fillUseForAlgEff();  // Fill useForAlgEff_ flag.
+
+    this->calcNumLayers();  // Calculate number of tracker layers this TP has stubs in.
+  }
+
+  //=== Check if this tracking particle is worth keeping.
+  //=== (i.e. If there is the slightest chance of reconstructing it, so as to measure fake rate).
+
+  void TP::fillUse() {
+    const bool useOnlyInTimeParticles = false;
+    const bool useOnlyTPfromPhysicsCollisionFalse = false;
+    // Use looser cuts here those those used for tracking efficiency measurement.
+    // Keep only those TP that have a chance (allowing for finite track resolution) of being reconstructed as L1 tracks. L1 tracks not matching these TP will be defined as fake.
+
+    const vector<int> genPdgIdsAllUnsigned = {
+        11,
+        13,
+        211,
+        321,
+        2212};  // Include all possible particle types here, as if some are left out, L1 tracks matching one of missing types will be declared fake.
+    vector<int> genPdgIdsAll;
+    for (unsigned int i = 0; i < genPdgIdsAllUnsigned.size(); i++) {
+      genPdgIdsAll.push_back(genPdgIdsAllUnsigned[i]);
+      genPdgIdsAll.push_back(-genPdgIdsAllUnsigned[i]);
+    }
+
+    // Range big enough to include all TP needed to measure tracking efficiency
+    // and big enough to include any TP that might be reconstructed for fake rate measurement.
+    const float ptMin = min(settings_->genMinPt(), 0.7 * settings_->houghMinPt());
+    const float etaMax = max(settings_->genMaxAbsEta(), 0.2 + fabs(settings_->etaRegions()[0]));
+
+    static thread_local TrackingParticleSelector trackingParticleSelector(ptMin,
```

Comment:
Replaced by "static const"

Reply:
Is any of the `TrackingParticleSelector` constructor parameters coming from configuration (the use of `Settings` would hint towards that)? If yes, I advice against `static` because if anyone would use a different instance of `Settings` with different values, a `static const trackingParticeSelector` is initialized with one of the settings (whichever gets "first"), which may lead to surprising results (that may be subtle or not).